src/common/ad_tx.c

```diff
@@ -268,42 +327,47 @@ ad_tx_assign(struct ad_tx *tx, void *addr, daos_size_t size, uint32_t val)
 /**
  * memset a storage region, save the operation for redo (and old value for undo if it's
  * required by @save_old).
+ * If AD_TX_LOG_ONLY is set for @flags, this function only logs the operation itself,
+ * it does not call the memset(), this is for reserve() interface.
  */
 int
-ad_tx_set(struct ad_tx *tx, void *addr, char c, daos_size_t size, bool save_old)
+ad_tx_set(struct ad_tx *tx, void *addr, char c, daos_size_t size, uint32_t flags)
 {
 	struct umem_act_item	*it_undo, *it_redo;
 
 	if (addr == NULL || size == 0 || size > UMEM_ACT_PAYLOAD_MAX_LEN)
 		return -DER_INVAL;
 
 	if (tx == NULL) {
-		memset(addr, c, size);
+		if (!(flags & AD_TX_LOG_ONLY))
+			memset(addr, c, size);
 		return 0;
 	}
 
-	D_ALLOC_ACT(it_redo, UMEM_ACT_SET, size);
-	if (it_redo == NULL)
-		return -DER_NOMEM;
-
-	if (save_old) {
+	if (flags & AD_TX_UNDO) {
 		D_ALLOC_ACT(it_undo, UMEM_ACT_COPY, size);
-		if (it_undo == NULL) {
-			D_FREE(it_redo);
+		if (it_undo == NULL)
 			return -DER_NOMEM;
-		}
 
 		memcpy(it_undo->it_act.ac_copy.payload, addr, size);
-		it_undo->it_act.ac_copy.addr = ad_ptr2addr(addr);
+		it_undo->it_act.ac_copy.addr = blob_ptr2addr(tx->tx_blob, addr);
 		it_undo->it_act.ac_copy.size = size;
 		AD_TX_ACT_ADD(tx, it_undo, ACT_UNDO);
 	}
 
-	it_redo->it_act.ac_set.addr = ad_ptr2addr(addr);
-	it_redo->it_act.ac_set.size = size;
-	it_redo->it_act.ac_set.val = c;
-	AD_TX_ACT_ADD(tx, it_redo, ACT_REDO);
+	if (!(flags & AD_TX_LOG_ONLY))
+		memset(addr, c, size);
+
+	if (flags & AD_TX_REDO) {
+		D_ALLOC_ACT(it_redo, UMEM_ACT_SET, size);
+		if (it_redo == NULL)
```

Comment:
Do we need free above memory if AD_TX_UNDO is set, or  cleanup will be done in the tx abort?

Reply:
No we don't need to, tx_end() will handle all these (even it doesn't, it should)